### PR TITLE
refactor: adjust outermost error message

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Adjust the user-facing error message to make it precise and concrete.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/client/src/database.rs [hunk 1]
```diff
@@ -113,15 +115,16 @@ impl Database {
...
-                    .unwrap_or(e.code() as u32);
-                let err_msg = get_metadata_value(&e, INNER_ERROR_MSG).unwrap_or(e.to_string());
-                error::FlightGetSnafu {
-                    addr: client.addr(),
-                    code,
-                    err_msg,
-                }
-                .build()
+                    .and_then(|s| StatusCode::from_str(&s).ok())
+                    .unwrap_or(StatusCode::Unknown);
```
Comment (MichaelScofield) [inline src/client/src/database.rs:119]: Tonic's error code (`e.code()`) is ignored if INNER_ERROR_CODE is not set.

File: src/client/src/error.rs [hunk 2]
```diff
@@ -25,18 +25,8 @@ pub enum Error {
...
-        addr,
-        code,
-        err_msg
-    ))]
-    FlightGet {
-        addr: String,
-        code: u32,
-        err_msg: String,
-        backtrace: Backtrace,
-    },
```
Comment (waynexia) [inline src/client/src/error.rs:39]: I split this error into two part. The new `FlightGet` describes an rpc call fail, with root cause and server address. `ExternalError` is a temporary error container for assembling errors from flight metadata. And the call stack of both is clear so I remove the backtrace.

**Review by evenyag — APPROVED** [review]

File: src/servers/src/error.rs [hunk 3]
```diff
@@ -347,11 +347,11 @@ impl From<Error> for tonic::Status {
 
         // If either of the status_code or error msg cannot convert to valid HTTP header value
         // (which is a very rare case), just ignore. Client will use Tonic status code and message.
-        if let Ok(code) = HeaderValue::from_bytes((err.status_code() as u32).to_string().as_bytes())
-        {
+        if let Ok(code) = HeaderValue::from_bytes(err.status_code().to_string().as_bytes()) {
             headers.insert(INNER_ERROR_CODE, code);
         }
-        if let Ok(err_msg) = HeaderValue::from_bytes(err.to_string().as_bytes()) {
+        let root_error = err.iter_chain().last().unwrap();
```
Comment (MichaelScofield) [inline src/servers/src/error.rs:353]: I think it's better not skip all the intermediate errors' messages

File: src/client/src/database.rs (hunk 1 above)
Comment (waynexia) [reply src/client/src/database.rs:119]: If we reach this step mean tonic rpc is success? Because we can read messages from server's flight metadata.

Comment (MichaelScofield) [reply src/client/src/database.rs:119]: not necessarily. better keep the tonic's code

File: src/servers/src/error.rs (hunk 3 above)
Comment (MichaelScofield) [reply src/servers/src/error.rs:353]: not good indeed

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**